File: iAsync.ObjC/lib/JFFUI/UIImage/UIImage+ScaleToSize.m
#import "UIImage+ScaleToSize.h"

@implementation UIImage (ScaleToSize)

- (UIImage *)imageScaledToSize:(CGSize)targetSize
                   contentMode:(UIViewContentMode)contentMode
{
    //other contentModes does not supported yet
    NSParameterAssert(UIViewContentModeScaleAspectFill == contentMode);
    
    CGSize originalSize = self.size;
    CGFloat scale = MAX(targetSize.width / originalSize.width, targetSize.height / originalSize.height);
    
    CGRect thumbnailRect = CGRectZero;
    thumbnailRect.size.width  = originalSize.width  * scale;
    thumbnailRect.size.height = originalSize.height * scale;
    
    UIGraphicsBeginImageContextWithOptions(thumbnailRect.size, YES, 0.0);
    
    [self drawInRect:thumbnailRect];
    
    UIImage *newImage = UIGraphicsGetImageFromCurrentImageContext();
    UIGraphicsEndImageContext();
    
    if (newImage == nil) NSLog(@"could not scale image");
    
    return newImage;
}

- (instancetype)imageScale:(CGFloat)scale
{
    UIGraphicsBeginImageContext(self.size);
    CGContextRef currentContext = UIGraphicsGetCurrentContext();
    
    CGContextTranslateCTM(currentContext, self.size.width/2.f, self.size.height/2.f);
    
    CGContextScaleCTM(currentContext, scale, -scale);
    
    CGRect drawRect = CGRectMake(-self.size.width/2.f, -self.size.height/2.f, self.size.width, self.size.height);
    CGContextDrawImage(currentContext, drawRect, self.CGImage);
    UIImage *cropped = UIGraphicsGetImageFromCurrentImageContext();
    UIGraphicsEndImageContext();
    
    return cropped;
}

@end

File: iAsync.Swift/lib/JUI/JUI/UIImage/UIImage+ScaleToSize.swift
//
//  UIImage+ScaleToSize.swift
//  JUI
//
//  Created by Vladimir Gorbenko on 19.03.15.
//  Copyright (c) 2015 EmbeddedSources. All rights reserved.
//

import Foundation

public extension UIImage {
    
    func imageScale(scale: CGFloat) -> UIImage {
        
        UIGraphicsBeginImageContext(self.size)
        let currentContext = UIGraphicsGetCurrentContext()
        
        CGContextTranslateCTM(currentContext, self.size.width/2.0, self.size.height/2.0)
        
        CGContextScaleCTM(currentContext, scale, -scale)
        
        let drawRect = CGRectMake(-self.size.width/2.0, -self.size.height/2.0, self.size.width, self.size.height)
        CGContextDrawImage(currentContext, drawRect, self.CGImage)
        let cropped = UIGraphicsGetImageFromCurrentImageContext()
        UIGraphicsEndImageContext()
        
        return cropped
    }
}
